scheduler/rank_test.go

```diff
@@ -2079,6 +2079,114 @@ func TestBinPackIterator_Device_Failure_With_Eviction(t *testing.T) {
 	must.Eq(t, 1, ctx.metrics.DimensionExhausted["devices: no devices match request"])
 }
 
+// Tests that bin packing iterator will not place workloads on nodes
+// that would go over a designated MaxAlloc value
+func TestBinPackIterator_MaxAlloc(t *testing.T) {
+	_, ctx := testContext(t)
+
+	taskGen := func(name string) *structs.Task {
+		return &structs.Task{
+			Name:      name,
+			Resources: &structs.Resources{},
+		}
+	}
+	nodes := []*RankedNode{
+		{
+			Node: &structs.Node{
+				ID: uuid.Generate(),
+				NodeResources: &structs.NodeResources{
+					Processors: processorResources2048,
+					Cpu:        legacyCpuResources2048,
+					Memory: structs.NodeMemoryResources{
+						MemoryMB: 2048,
+					},
+				},
+			},
+		},
+		{
+			Node: &structs.Node{
+				ID: uuid.Generate(),
+				NodeResources: &structs.NodeResources{
+					Processors: processorResources2048,
+					Cpu:        legacyCpuResources2048,
+					Memory: structs.NodeMemoryResources{
+						MemoryMB: 2048,
+					},
+				},
+			},
+		},
+	}
+
+	testCases := []struct {
+		name        string
+		maxAlloc    int
+		tasks       []*structs.Task
+		tasksOn1    int
+		nodesPlaced int
+		noNodes     bool
+	}{
+		{
+			name:     "both_nodes",
+			maxAlloc: 1,
+			tasks: []*structs.Task{
+				taskGen("web1"),
+				taskGen("web2"),
+				taskGen("web3")},
+			nodesPlaced: 2,
+		},
+		{
+			name:     "only_node2",
+			maxAlloc: 0,
+			tasks: []*structs.Task{
+				taskGen("web1"),
+				taskGen("web2"),
+				taskGen("web3")},
+			nodesPlaced: 1,
+		},
+		{
+			name:     "no_nodes",
+			maxAlloc: 0,
+			tasks: []*structs.Task{
+				taskGen("web1"),
+				taskGen("web2"),
+				taskGen("web3")},
+			nodesPlaced: 0,
+			noNodes:     true,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.noNodes {
+				// Note: test case order matters here
+				for _, v := range nodes {
+					v.Node.NodeMaxAllocs = &structs.NodeMaxAllocs{
+						MaxAllocs: tc.maxAlloc,
+					}
+				}
+			}
+			// only add allocation limit to first node except in noNodes
+			nodes[0].Node.NodeMaxAllocs = &structs.NodeMaxAllocs{
+				MaxAllocs: tc.maxAlloc,
+			}
```

Comment:
This test is fine in terms of correctness. But from a test design standpoint it might be nice if you had explicit `maxAllocNode0` and `maxAllocNode1` fields in the test cases instead of having conditional test setup logic here. That way someone reading the test can look at a given test case and just see the inputs and outputs, and the test itself has no logic except calling the functions under test.